tests/drivers/bc12/src/main.c

```diff
@@ -0,0 +1,196 @@
+		fixture->charging_partner.type = BC12_TYPE_NONE;
+		fixture->charging_partner.current = 0;
+		fixture->charging_partner.voltage = 0;
+	}
+}
+
+ZTEST_USER_F(bc12_pi3usb9201, test_bc12_no_charging_partner)
+{
+	bc12_result_cb(fixture->bc12_dev, &bc12_test_result_cb, fixture);
+
+	pi3usb9201_emul_set_charging_partner(fixture->bc12_emul, BC12_TYPE_NONE);
+
+	bc12_set_role(fixture->bc12_dev, BC12_PORTABLE_DEVICE);
+
+	k_sleep(K_MSEC(100));
+
+	/* Without any device connected, our callback should not execute */
+	zassert_equal(fixture->callback_count, 0);
+}
+
+ZTEST_USER_F(bc12_pi3usb9201, test_bc12_sdp_charging_partner)
+{
+	bc12_result_cb(fixture->bc12_dev, &bc12_test_result_cb, fixture);
+
+	/* Connect a SDP charging partner to the emulator */
+	pi3usb9201_emul_set_charging_partner(fixture->bc12_emul, BC12_TYPE_SDP);
+
+	/* Report to the BC1.2 driver that VBUS is present */
+	bc12_set_role(fixture->bc12_dev, BC12_PORTABLE_DEVICE);
+
+	k_sleep(K_MSEC(100));
+
+	zassert_equal(fixture->callback_count, 1);
+	zassert_equal(fixture->charging_partner.type, BC12_TYPE_SDP);
+	zassert_equal(fixture->charging_partner.current, 500);
+	zassert_equal(fixture->charging_partner.voltage, 5000);
+
+	/* Remove the charging partner */
+	fixture->callback_count = 0;
+	pi3usb9201_emul_set_charging_partner(fixture->bc12_emul, BC12_TYPE_NONE);
+
+	/* Report to the BC1.2 driver that VBUS is no longer present */
+	bc12_set_role(fixture->bc12_dev, BC12_DISCONNECTED);
+
+	k_sleep(K_MSEC(100));
+
+	/* The BC1.2 driver should invoke the callback on disconnects */
+	zassert_equal(fixture->callback_count, 1);
+	zassert_equal(fixture->charging_partner.type, BC12_TYPE_NONE);
+	zassert_equal(fixture->charging_partner.current, 0);
+	zassert_equal(fixture->charging_partner.voltage, 0);
+}
+
+ZTEST_USER_F(bc12_pi3usb9201, test_bc12_cdp_charging_partner)
+{
+	bc12_result_cb(fixture->bc12_dev, &bc12_test_result_cb, fixture);
+
+	pi3usb9201_emul_set_charging_partner(fixture->bc12_emul, BC12_TYPE_CDP);
+
+	bc12_set_role(fixture->bc12_dev, BC12_PORTABLE_DEVICE);
+
+	k_sleep(K_MSEC(100));
+
+	zassert_equal(fixture->callback_count, 1);
+	zassert_equal(fixture->charging_partner.type, BC12_TYPE_CDP);
+	zassert_equal(fixture->charging_partner.current, 1500);
+	zassert_equal(fixture->charging_partner.voltage, 5000);
+
+	/* Remove the charging partner */
+	fixture->callback_count = 0;
+	pi3usb9201_emul_set_charging_partner(fixture->bc12_emul, BC12_TYPE_NONE);
+
+	/* Report to the BC1.2 driver that VBUS is no longer present */
+	bc12_set_role(fixture->bc12_dev, BC12_DISCONNECTED);
+
+	k_sleep(K_MSEC(100));
+
+	/* The BC1.2 driver should invoke the callback on disconnects */
+	zassert_equal(fixture->callback_count, 1);
+	zassert_equal(fixture->charging_partner.type, BC12_TYPE_NONE);
+	zassert_equal(fixture->charging_partner.current, 0);
+	zassert_equal(fixture->charging_partner.voltage, 0);
+}
+
+ZTEST_USER_F(bc12_pi3usb9201, test_bc12_sdp_to_cdp_charging_partner)
```

Comment:
Am I mixing up all the confusing BC1.2 abbreviations, or did you typo this function name?

Reply:
Thanks, fixed.